docs/guides/implementation-guides/how-to-implement-dependency-injection.md

```diff
@@ -3,13 +3,142 @@ id: how-to-implement-dependency-injection
+
+Dependency injection solves these problem by abstracting away the creation of objects and their dependencies. This allows for well encapsulated services to be used that will be automatically passed into any other service that is registered to use them. 
+
+See below the steps you need to do in order to resolve that dependency using DI.
+
+## Step 1: Install the NuGet package for DI
+There are many dependency injection (DI) container providers available ([DryIoC](https://github.com/dadhi/DryIoc), [Autofac](https://github.com/autofac/Autofac), [Pure.DI](https://github.com/DevTeam/Pure.DI)) but this guide will only focus on `Microsoft.Extensions.DependencyInjection` which is a lightweight, extensible dependency injection container. It provides an easy-to-use and convention-based way to add DI to .NET applications, including Avalonia-based desktop applications.
+
+```shell
+dotnet add package Microsoft.Extensions.DependencyInjection
+```
+
+## Step 2: Add ServiceCollectionExtensions 
+The following code is creating an extension for IServiceCollection. We are creating different methods to be use in different settings (Design mode and real usage). You usually want to mock services in design mode with the library of your choice.    
+
+```csharp
+public static class ServiceCollectionExtensions {
+
+    public static void RegisterDesignTimeServices(this IServiceCollection collection)
+    {
+        collection.AddTransient<MockRepository>(); // MockRepository is a mock instance of IRepository
+    }
+
+    public static void RegisterRealServices(this IServiceCollection collection)
+    {
+        collection.AddTransient<Repository>();
+        collection.AddTransient<BusinessService>();
+    }
+
+    public static void AddCommonServices(this IServiceCollection collection) {
+        collection.AddTransient<MainViewModel>();
+    }
+}
+```
+
+## Step 3: Modify App.axaml.cs
+Next, the App.xaml.cs class should be modified to use the DI container. This will allow the previously registered view model to be resolved via the dependency injection container. The fully realised view model can then be set to the data context of the main view. 
+
+```csharp
+public partial class App : Application
+{
+    public override void Initialize()
+    {
+        AvaloniaXamlLoader.Load(this);
+    }
+
+    public override void OnFrameworkInitializationCompleted()
+    {
+        // Line below is needed to remove Avalonia data validation.
+        // Without this line you will get duplicate validations from both Avalonia and CT
+        BindingPlugins.DataValidators.RemoveAt(0);
+
+        // Register all the services needed for the application to run
+        // If in design mode we register different set of services using the extension method from step 2
+        var collection = new ServiceCollection();
+        collection.AddCommonServices();
+        if (Design.IsDesignMode)
```

Comment:
I should mention: Since you decided to go with the whole design/real services from my example, you can't use constructor injection with the previewer, at least out of the box. It wont work in most cases. You need to use a markup extension to resolve the services and set the design time data context. I excluded it from my example to avoid overcomplicating it since it also brings back in the locator as there is no way to pass a reference to the container into the markup extension that I know of. I was fairly lazy here so I just shoved the locator into the App class. But that service manager class would also work pretty well here.

```csharp
    public class DesignLocator : MarkupExtension {
        #region Public properties
        public Type Type { get; init; }
        #endregion

        #region Public methods
        public override object ProvideValue(IServiceProvider serviceProvider) {
            if (!Design.IsDesignMode) {
                throw new InvalidOperationException();
            }

            return App.Locator.GetRequiredService(this.Type);
        }
        #endregion
    }
```

You use it like this:

```xaml
        Design.DataContext="{markupExtensions:DesignLocator Type={x:Type vm:MainWindowViewModel}}"
```

See https://github.com/AvaloniaUI/Avalonia/discussions/13743 for the source for this idea.

Reply:
I'm a bit uncertain on how to test this on my local sample, so I prefer to remove the design part. It's probably another guide to create in the documentation.